Native updater fails on latest upstream DMG: critical patch failures

## Summary

The native updater detects the latest upstream Codex DMG and starts a local rebuild, but install.sh aborts during the Linux patching phase because several required patches no longer match the current upstream bundle.

This also reproduces when building against the same latest DMG manually, so it looks like an upstream bundle/layout change rather than a pacman/native packaging issue.

## Environment

- Distro: Arch Linux x86_64
- Kernel: 7.0.9-arch2-1
- Desktop/session: Hyprland / Wayland
- Package format: pacman
- Installed package: codex-desktop 2026.06.26.015509-1
- Wrapper repo commit: 407ff0e31cddd64a8c671d1351e962a07e6cb6c9 (Better proxy env handling #565)
- Detected Electron version: 42.1.0
- Codex CLI: 0.142.2 / up to date

## Updater status

~~~json
{
  "installed_version": "2026.06.26.015509-1",
  "candidate_version": "2026.06.26.015900+54a24070",
  "status": "failed",
  "remote_headers_fingerprint": "etag=0x8DED2E449ADFFF8|last_modified=Thu, 25 Jun 2026 18:05:06 GMT|content_length=529265008",
  "dmg_sha256": "54a2407090f91194ec7339912e071d2360e557cf631eb99e0aae7bd5c394857f",
  "error_message": "install.sh failed during local rebuild",
  "auto_install_on_app_exit": true,
  "waiting_for_app_exit_auto_install": false,
  "cli_status": "up_to_date"
}
~~~

## install.log excerpt

~~~text
[INFO] Using provided DMG: ~/.cache/codex-update-manager/downloads/Codex.dmg
[INFO] Extracting DMG with 7z...
[INFO] 7z reported 9 safe package symlink warnings; repaired and continuing
[INFO] Found: Codex.app
[INFO] Detected Electron version from package.json: 42.1.0
[INFO] Extracting app.asar...
[INFO] Native modules: better-sqlite3@12.9.0, node-pty@1.1.0
[INFO] Installing fresh sources from npm...
[INFO] Patched better-sqlite3 source for V8 external pointer API
[INFO] Compiling for Electron v42.1.0 (this takes ~1 min)...
[INFO] Using Electron headers: https://artifacts.electronjs.org/headers/dist
Searching dependency tree
Building modules: better-sqlite3, node-pty
✔ Rebuild Complete
[INFO] Native modules built successfully
[INFO] Patching Linux window and shell behavior...
WARN: Could not find avatar overlay drag start — skipping Linux avatar overlay drag interactivity patch
WARN: Could not find avatar overlay drag end — skipping Linux avatar overlay drag cleanup patch
WARN: Could not find avatar overlay element size update — skipping Linux avatar overlay layout interactivity patch
WARN: Could not find avatar overlay layout application — skipping Linux avatar overlay layout sync patch
WARN: Could not find avatar overlay show window — skipping Linux avatar overlay show sync patch
WARN: Could not find Browser Use node_repl config insertion point — skipping node_repl approval patch
WARN: Could not find Owl feature binding loader - skipping Linux Owl feature fallback patch
WARN: Could not find automation schedule bundle in /tmp/tmp.CDT2LhTarA/app-extracted/.vite/build or /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping RRULE multi-time patch
WARN: Could not find browser annotation screenshot markers — skipping screenshot marker patch
WARN: Could not find resolved theme bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping translucent sidebar default patch
WARN: Could not find fast-mode/service-tier availability bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping fast-mode model guard patch
WARN: Could not find window controls safe-area bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping Linux window controls safe-area patch
WARN: Could not find profile dropdown webview bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping Linux profile settings menu patch
WARN: Could not find font settings bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping Linux monospace font stack patch
WARN: Could not find subagent metadata webview bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping subagent nickname metadata shape patch
WARN: Could not find tooltip bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping Linux tooltip titlebar collision patch
WARN: Could not find Browser Use availability bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping Linux Browser Use availability patch
WARN: Could not find external Browser Use availability bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping Linux external Browser Use availability patch
WARN: Keybinds settings patch skipped: Required Keybinds settings patch failed: could not find JSX runtime asset
Critical patch failures (5):
  - linux-avatar-overlay-mouse-passthrough (failed-required): WARN: Could not find avatar overlay drag start — skipping Linux avatar overlay drag interactivity patch
  - main-process-ui (failed-required): WARN: Could not find avatar overlay drag start — skipping Linux avatar overlay drag interactivity patch
  - linux-owl-feature-binding-fallback (failed-required): WARN: Could not find Owl feature binding loader - skipping Linux Owl feature fallback patch
  - linux-fast-mode-model-guard (failed-required): WARN: Could not find fast-mode/service-tier availability bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping fast-mode model guard patch
  - subagent-nickname-metadata-shape (failed-required): WARN: Could not find subagent metadata webview bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping subagent nickname metadata shape patch
Aborting: these patches are required for a working Linux app. Set CODEX_ENFORCE_CRITICAL_PATCHES=0 to bypass (emergency builds only).
[ERROR] Failed at line 99 (exit code 1)
~~~

## Failed required patches from patch-report.json

~~~json
[
  {
    "name": "linux-avatar-overlay-mouse-passthrough",
    "status": "failed-required",
    "reason": "WARN: Could not find avatar overlay drag start — skipping Linux avatar overlay drag interactivity patch"
  },
  {
    "name": "main-process-ui",
    "status": "failed-required",
    "reason": "WARN: Could not find avatar overlay drag start — skipping Linux avatar overlay drag interactivity patch"
  },
  {
    "name": "linux-owl-feature-binding-fallback",
    "status": "failed-required",
    "reason": "WARN: Could not find Owl feature binding loader - skipping Linux Owl feature fallback patch"
  },
  {
    "name": "linux-fast-mode-model-guard",
    "status": "failed-required",
    "reason": "WARN: Could not find fast-mode/service-tier availability bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping fast-mode model guard patch"
  },
  {
    "name": "subagent-nickname-metadata-shape",
    "status": "failed-required",
    "reason": "WARN: Could not find subagent metadata webview bundle in /tmp/tmp.CDT2LhTarA/app-extracted/webview/assets — skipping subagent nickname metadata shape patch"
  }
]
~~~

## Expected behavior

The updater/build should either successfully patch/package the latest upstream DMG, or clearly identify that the current upstream Codex build is temporarily unsupported.

## Actual behavior

The updater detects the candidate, rebuilds native modules successfully, then aborts during required Linux patching. Existing installed app remains usable, but automatic native update cannot apply this candidate.

--- a/scripts/patches/automation-schedule.js
+++ b/scripts/patches/automation-schedule.js
@@ -33,6 +33,10 @@ function findWorkspaceRootDropHandlerBundles(extractedDir) {
             /^automation-schedule-.*\.js$/.test(path.basename(candidate)) &&
             source.includes("hasMultipleTimeValues") &&
             source.includes("function Tn(")
+          ) ||
+          (
+            source.includes("hasMultipleTimeValues") &&
+            /rruleText:e,time:\w+\(r\.byhour,r\.byminute,r\)/.test(source)
           );
       } catch {
         return false;
@@ -115,11 +119,77 @@ function applyAutomationScheduleMultiTimePatch(source) {
     if (currentPatched !== source) {
       return currentPatched;
     }
+    const genericPatched = applyGenericAutomationScheduleMultiTimePatch(source);
+    if (genericPatched !== source) {
+      return genericPatched;
+    }
     console.warn("WARN: Could not find automation schedule helper block — skipping RRULE multi-time patch");
     return source;
   }
 
   return source.slice(0, block.start) + automationScheduleReplacement(block) + source.slice(block.end);
+}
+
+function applyGenericAutomationScheduleMultiTimePatch(source) {
+  if (source.includes("function codexLinuxNormalizeRruleNumbers(")) {
+    return source;
+  }
+
+  const parserRe =
+    /hasMultipleTimeValues:Array\.isArray\(r\.byhour\)&&r\.byhour\.length>1\|\|Array\.isArray\(r\.byminute\)&&r\.byminute\.length>1,interval:Math\.max\(1,Math\.round\(r\.interval\?\?1\)\),minute:a,origOptions:n\.origOptions,rruleText:e,time:(\w+)\(r\.byhour,r\.byminute,r\),weekdays:i/;
+  const parserMatch = parserRe.exec(source);
+  if (!parserMatch) {
+    return source;
+  }
+  const timeFn = parserMatch[1];
+
+  const helperRe = new RegExp(
+    "function " +
+      timeFn +
+      "\\(e,t,n\\)\\{let r=(\\w+)\\(e\\),i=\\1\\(t\\);return r!=null&&i!=null\\?(\\w+)\\(r,i\\):n\\.dtstart\\?\\2\\(n\\.dtstart\\.getHours\\(\\),n\\.dtstart\\.getMinutes\\(\\)\\):(\\w+)\\}function \\1\\(e\\)\\{return Array\\.isArray\\(e\\)\\?typeof e\\[0\\]==`number`\\?e\\[0\\]:null:typeof e==`number`\\?e:null\\}",
+  );
+  const helperMatch = helperRe.exec(source);
+  if (!helperMatch) {
+    return source;
+  }
+  const helperBlock = helperMatch[0];
+  const combineFn = helperMatch[2];
+
+  const summaryRe =
+    /function (\w+)\(e,t\)\{if\(!e\|\|e\.hasMultipleTimeValues\)return null;[\s\S]*?let i=(\w+)\(e\.time,t\);return i\?(\w+)\(\{intl:t,isEveryDay:r,timeLabel:i,weekdays:n\}\):null\}/;
+  const summaryMatch = summaryRe.exec(source);
+  if (!summaryMatch) {
+    return source;
+  }
+  const summaryBlock = summaryMatch[0];
+  const labelFn = summaryMatch[2];
+
+  const helperPatch =
+    helperBlock +
+    "function codexLinuxNormalizeRruleNumbers(e,t,n){let r=Array.isArray(e)?e:[e];return Array.from(new Set(r.filter(e=>typeof e==`number`&&Number.isInteger(e)&&e>=t&&e<=n))).sort((e,t)=>e-t)}" +
+    "function codexLinuxRruleTimes(e,t,n){let r=codexLinuxNormalizeRruleNumbers(e,0,23),i=codexLinuxNormalizeRruleNumbers(t,0,59);n.dtstart&&(r.length!==0||(r=[n.dtstart.getHours()]),i.length!==0||(i=[n.dtstart.getMinutes()]));let a=[];for(let e of r)for(let t of i)a.push(" +
+    combineFn +
+    "(e,t));return Array.from(new Set(a)).sort()}" +
+    "function codexLinuxAutomationTimeLabel(e,t){let n=Array.isArray(e.timeValues)&&e.timeValues.length>0?e.timeValues:[e.time],r=n.map(e=>" +
+    labelFn +
+    "(e,t)).filter(Boolean);return r.length===0?null:typeof t.formatList==`function`?t.formatList(r,{type:`conjunction`}):r.join(`, `)}";
+
+  const parserPatch =
+    "hasMultipleTimeValues:codexLinuxRruleTimes(r.byhour,r.byminute,r).length>1,interval:Math.max(1,Math.round(r.interval??1)),minute:a,origOptions:n.origOptions,rruleText:e,time:" +
+    timeFn +
+    "(r.byhour,r.byminute,r),timeValues:codexLinuxRruleTimes(r.byhour,r.byminute,r),weekdays:i";
+
+  const summaryPatch = summaryBlock
+    .replace("if(!e||e.hasMultipleTimeValues)return null;", "if(!e)return null;")
+    .replace(
+      "let i=" + labelFn + "(e.time,t);",
+      "let i=codexLinuxAutomationTimeLabel(e,t);",
+    );
+
+  let patched = source.replace(helperBlock, () => helperPatch);
+  patched = patched.replace(parserMatch[0], () => parserPatch);
+  patched = patched.replace(summaryBlock, () => summaryPatch);
+  return patched;
 }
 
 function applyCurrentAutomationScheduleMultiTimePatch(source) {

--- a/scripts/patches/avatar-overlay.js
+++ b/scripts/patches/avatar-overlay.js
@@ -60,7 +60,11 @@ function findAvatarOverlayClass(source) {
     const openIndex = match.index + match[0].length - 1;
     const closeIndex = findMatchingBrace(source, openIndex);
     if (closeIndex === -1) {
-      return null;
+      // findMatchingBrace cannot balance this class (e.g. it contains a regex
+      // literal with an unbalanced brace). Skip it and keep scanning instead of
+      // aborting — the avatar overlay class may appear later in the bundle.
+      classRegex.lastIndex = openIndex + 1;
+      continue;
     }
     const text = source.slice(match.index, closeIndex + 1);
     if (

--- a/scripts/patches/core/all-linux/main-process/browser-integrations/patch.js
+++ b/scripts/patches/core/all-linux/main-process/browser-integrations/patch.js
@@ -1,7 +1,8 @@
 "use strict";
 
+const { patchStatusFromChange } = require("../../../../../lib/patch-report.js");
 const {
-  applyBrowserUseNodeReplApprovalPatch,
+  applyBrowserUseNodeReplApprovalAssets,
   applyLinuxBrowserUseRouteLivenessPatch,
   applyLinuxChromeExtensionStatusPatch,
 } = require("../../../../main-process.js");
@@ -17,10 +18,20 @@ module.exports = [
   },
   {
     id: "browser-use-node-repl-approval",
-    phase: "main-bundle",
+    phase: "extracted-app",
     order: 160,
     ciPolicy: "optional",
-    apply: applyBrowserUseNodeReplApprovalPatch,
+    apply: applyBrowserUseNodeReplApprovalAssets,
+    status: (result, warnings) => ({
+      status:
+        result?.matched === 0
+          ? "skipped-optional"
+          : patchStatusFromChange(Boolean(result?.changed), warnings, "optional"),
+      reason:
+        result?.matched === 0
+          ? "Browser Use node_repl mcp config bundle not found"
+          : warnings[0] ?? null,
+    }),
   },
   {
     id: "linux-browser-use-route-liveness",

--- a/scripts/patches/core/all-linux/webview/browser-use-availability/patch.js
+++ b/scripts/patches/core/all-linux/webview/browser-use-availability/patch.js
@@ -9,7 +9,7 @@ module.exports = {
   phase: "webview-asset",
   order: 1090,
   ciPolicy: "optional",
-  pattern: /^(use-in-app-browser-use-availability|use-is-plugins-enabled)-.*\.js$/,
+  pattern: /^(?:use-in-app-browser-use-availability-|use-is-plugins-enabled-|app-initial~app-main~).*\.js$/,
   missingDescription: "Browser Use availability bundle",
   skipDescription: "Linux Browser Use availability patch",
   apply: applyLinuxBrowserUseAvailabilityPatch,

--- a/scripts/patches/core/all-linux/webview/browser-use-external-availability/patch.js
+++ b/scripts/patches/core/all-linux/webview/browser-use-external-availability/patch.js
@@ -9,7 +9,7 @@ module.exports = {
   phase: "webview-asset",
   order: 1092,
   ciPolicy: "optional",
-  pattern: /^(use-is-plugins-enabled|use-in-app-browser-use-availability)-.*\.js$/,
+  pattern: /^(?:use-is-plugins-enabled|use-in-app-browser-use-availability)-.*\.js$|^app-initial~app-main~.*\.js$/,
   missingDescription: "external Browser Use availability bundle",
   skipDescription: "Linux external Browser Use availability patch",
   apply: applyLinuxBrowserUseExternalAvailabilityPatch,

--- a/scripts/patches/core/all-linux/webview/fast-mode-guard/patch.js
+++ b/scripts/patches/core/all-linux/webview/fast-mode-guard/patch.js
@@ -8,7 +8,10 @@ module.exports = [
     phase: "webview-asset",
     order: 1040,
     ciPolicy: "required-upstream",
-    pattern: /^(?:use-is-fast-mode-enabled|read-service-tier-for-request|use-service-tier-settings|app-server-manager-signals)-.*\.js$/,
+    // Older DMGs emit granular hook chunks; 26.623+ merges service-tier code
+    // into the shared `app-initial~app-main~…` bundle (and switched to optional
+    // chaining, so the guard is a no-op there). Match both shapes.
+    pattern: /^(?:use-is-fast-mode-enabled|read-service-tier-for-request|use-service-tier-settings|app-server-manager-signals|app-initial~app-main~).*\.js$/,
     missingDescription: "fast-mode/service-tier availability bundle",
     skipDescription: "fast-mode model guard patch",
     apply: applyLinuxFastModeModelGuardPatch,

--- a/scripts/patches/core/all-linux/webview/font-settings/patch.js
+++ b/scripts/patches/core/all-linux/webview/font-settings/patch.js
@@ -8,7 +8,7 @@ module.exports = [
     phase: "webview-asset",
     order: 1045,
     ciPolicy: "optional",
-    pattern: /^font-settings-.*\.js$/,
+    pattern: /^(?:font-settings-.*|app-initial~app-main~remote-conversation-page~.*)\.js$/,
     missingDescription: "font settings bundle",
     skipDescription: "Linux monospace font stack patch",
     apply: applyLinuxSafeMonospaceFontStackPatch,

--- a/scripts/patches/core/all-linux/webview/profile-settings-menu/patch.js
+++ b/scripts/patches/core/all-linux/webview/profile-settings-menu/patch.js
@@ -8,7 +8,7 @@ module.exports = [
     phase: "webview-asset",
     order: 1043,
     ciPolicy: "optional",
-    pattern: /^profile-dropdown-.*\.js$/,
+    pattern: /^(?:profile-dropdown-.*|app-initial~app-main~automations-page-.*)\.js$/,
     missingDescription: "profile dropdown webview bundle",
     skipDescription: "Linux profile settings menu patch",
     apply: applyLinuxProfileSettingsMenuPatch,

--- a/scripts/patches/core/all-linux/webview/subagent-metadata/patch.js
+++ b/scripts/patches/core/all-linux/webview/subagent-metadata/patch.js
@@ -10,7 +10,10 @@ module.exports = [
     phase: "webview-asset",
     order: 1050,
     ciPolicy: "required-upstream",
-    pattern: /^(?:app-server-manager-signals|use-host-config)-.*\.js$/,
+    // Older DMGs emit granular hook chunks; 26.623+ merges them into the shared
+    // `app-initial~app-main~…` bundle. Match both so the patch keeps targeting
+    // the chunk that actually carries the subagent metadata module.
+    pattern: /^(?:app-server-manager-signals|use-host-config|app-initial~app-main~).*\.js$/,
     missingDescription: "subagent metadata webview bundle",
     skipDescription: "subagent nickname metadata shape patch",
     apply: applySubagentNicknameMetadataPatch,

--- a/scripts/patches/core/all-linux/webview/theme-and-sunset/patch.js
+++ b/scripts/patches/core/all-linux/webview/theme-and-sunset/patch.js
@@ -44,7 +44,7 @@ module.exports = [
     phase: "webview-asset",
     order: 1030,
     ciPolicy: "optional",
-    pattern: /^(diff-view-mode|use-resolved-theme-variant)-.*\.js$/,
+    pattern: /^(diff-view-mode|use-resolved-theme-variant)-.*\.js$|^app-initial~app-main~.*projects-index-page~app~.*\.js$/,
     missingDescription: "resolved theme bundle",
     skipDescription: "translucent sidebar default patch",
     apply: applyLinuxOpaqueWindowsDefaultPatch,
@@ -54,7 +54,7 @@ module.exports = [
     phase: "webview-asset",
     order: 1040,
     ciPolicy: "optional",
-    pattern: /^use-window-controls-safe-area-.*\.js$/,
+    pattern: /^(?:use-window-controls-safe-area-|app-initial~app-main~remote-conversation-page~new-thread-panel-page~projects-index-page~app~).*\.js$/,
     missingDescription: "window controls safe-area bundle",
     skipDescription: "Linux window controls safe-area patch",
     apply: applyLinuxWindowControlsSafeAreaPatch,
@@ -64,7 +64,10 @@ module.exports = [
     phase: "webview-asset",
     order: 1050,
     ciPolicy: "optional",
-    pattern: /^tooltip-.*\.js$/,
+    // 26.623 merged the tooltip floating-ui middleware into the shared
+    // `app-initial~app-main~…` bundle; keep matching the old granular
+    // `tooltip-*` chunk and add the merged bundle so the patch keeps landing.
+    pattern: /^(?:tooltip-|app-initial~app-main~).*\.js$/,
     missingDescription: "tooltip bundle",
     skipDescription: "Linux tooltip titlebar collision patch",
     apply: applyLinuxTooltipWindowControlsCollisionPatch,

--- a/scripts/patches/main-process/browser.js
+++ b/scripts/patches/main-process/browser.js
@@ -1,5 +1,8 @@
 "use strict";
 
+const fs = require("node:fs");
+const path = require("node:path");
+
 const {
   requireName,
 } = require("../shared.js");
@@ -41,6 +44,23 @@ function applyBrowserUseNodeReplApprovalPatch(currentSource) {
     (_match, runtimeFactoryVar, runtimeFactoryMethod, configPrefix, nodeReplPathVar, comma) => {
       patchedAnyCurrentRuntimeConfig = true;
       return `${runtimeFactoryVar}.${runtimeFactoryMethod}({${configPrefix}nodeReplPath:${nodeReplPathVar}${comma}tools:{js:{approval_mode:\`approve\`}},`;
+    },
+  );
+
+  // 26.623+: the browser-use node_repl mcp server config is emitted as a
+  // standalone object literal `{args:[],command:VAR,env:VAR,startup_timeout_sec:120}`
+  // (env is now a hoisted variable, not the old inline `env:{...}` literal) inside
+  // a separate `src-*.js` chunk. Insert the js auto-approval there.
+  const mcpServerConfigRegex =
+    /(\[`mcp_servers\.\$\{[A-Za-z_$][\w$]*\}`\]:\{args:\[\],command:[A-Za-z_$][\w$]*,env:[A-Za-z_$][\w$]*,)(startup_timeout_sec:120\})/g;
+  const mcpServerConfigAlreadyApprovedRegex =
+    /\[`mcp_servers\.\$\{[A-Za-z_$][\w$]*\}`\]:\{args:\[\],command:[A-Za-z_$][\w$]*,env:[A-Za-z_$][\w$]*,tools:\{js:\{approval_mode:`approve`\}\},startup_timeout_sec:120\}/;
+  let patchedAnyMcpServerConfig = false;
+  patchedSource = patchedSource.replace(
+    mcpServerConfigRegex,
+    (_match, configPrefix, configSuffix) => {
+      patchedAnyMcpServerConfig = true;
+      return `${configPrefix}tools:{js:{approval_mode:\`approve\`}},${configSuffix}`;
     },
   );
 
@@ -94,7 +114,8 @@ function applyBrowserUseNodeReplApprovalPatch(currentSource) {
   if (
     !patchedTrustedHashes &&
     !patchedSource.includes("codexLinuxTrustedBrowserClientSha256s(") &&
-    patchedSource.includes("NODE_REPL_TRUSTED_BROWSER_CLIENT_SHA256S")
+    patchedSource.includes("NODE_REPL_TRUSTED_BROWSER_CLIENT_SHA256S") &&
+    /trustedBrowserClientSha256s:[^,{}]+\|\|/.test(patchedSource)
   ) {
     console.warn(
       "WARN: Could not find Browser Use trusted hash insertion point — skipping Linux Browser Use trusted hash patch",
@@ -103,9 +124,12 @@ function applyBrowserUseNodeReplApprovalPatch(currentSource) {
 
   if (
     patchedSource === currentSource &&
+    patchedSource.includes("startup_timeout_sec:120") &&
     !patchedSource.includes(approvalPatch) &&
     !patchedAnyCurrentRuntimeConfig &&
     !currentRuntimeConfigAlreadyApprovedRegex.test(patchedSource) &&
+    !patchedAnyMcpServerConfig &&
+    !mcpServerConfigAlreadyApprovedRegex.test(patchedSource) &&
     !patchedTrustedHashes &&
     !patchedSource.includes("codexLinuxTrustedBrowserClientSha256s(")
   ) {
@@ -115,6 +139,50 @@ function applyBrowserUseNodeReplApprovalPatch(currentSource) {
   }
 
   return patchedSource;
+}
+
+// 26.623+: the node_repl mcp config (and the NODE_REPL_TRUSTED_BROWSER_CLIENT_SHA256S
+// trusted-hash code) was re-chunked out of the main bundle into a sibling
+// `.vite/build/src-*.js` chunk. Scan every build chunk that carries either marker so
+// the approval patch reaches whichever chunk now hosts the config.
+function applyBrowserUseNodeReplApprovalAssets(extractedDir) {
+  const buildDir = path.join(extractedDir, ".vite", "build");
+  if (!fs.existsSync(buildDir)) {
+    return { matched: 0, changed: 0 };
+  }
+
+  const candidates = fs
+    .readdirSync(buildDir)
+    .filter((name) => name.endsWith(".js"))
+    .sort()
+    .map((name) => path.join(buildDir, name))
+    .filter((candidate) => {
+      try {
+        const source = fs.readFileSync(candidate, "utf8");
+        return (
+          source.includes("startup_timeout_sec:120") ||
+          source.includes("NODE_REPL_TRUSTED_BROWSER_CLIENT_SHA256S")
+        );
+      } catch {
+        return false;
+      }
+    });
+
+  let changed = 0;
+  const pendingWrites = [];
+  for (const candidate of candidates) {
+    const currentSource = fs.readFileSync(candidate, "utf8");
+    const patchedSource = applyBrowserUseNodeReplApprovalPatch(currentSource);
+    if (patchedSource !== currentSource) {
+      changed += 1;
+      pendingWrites.push({ filePath: candidate, patchedSource });
+    }
+  }
+  for (const { filePath, patchedSource } of pendingWrites) {
+    fs.writeFileSync(filePath, patchedSource, "utf8");
+  }
+
+  return { matched: candidates.length, changed };
 }
 
 function applyLinuxBrowserUseRouteLivenessPatch(currentSource) {
@@ -252,6 +320,7 @@ function applyLinuxChromeExtensionStatusPatch(currentSource) {
 
 module.exports = {
   applyBrowserUseNodeReplApprovalPatch,
+  applyBrowserUseNodeReplApprovalAssets,
   applyLinuxBrowserUseRouteLivenessPatch,
   applyLinuxChromeExtensionStatusPatch,
 };

--- a/scripts/patches/main-process/misc.js
+++ b/scripts/patches/main-process/misc.js
@@ -176,6 +176,21 @@ function applyLinuxOwlFeatureBindingFallbackPatch(currentSource) {
     /function ([A-Za-z_$][\w$]*)\(\)\{let ([A-Za-z_$][\w$]*)=process\._linkedBinding;if\(typeof \2!=`function`\)throw Error\(`Owl feature binding is unavailable`\);return ([A-Za-z_$][\w$]*)\.parse\(\2\.call\(process,`electron_common_owl_features`\)\)\}/u;
   const match = currentSource.match(loaderRegex);
   if (match == null) {
+    // 26.623+ rewrote the loader to natively return null when the binding is
+    // unavailable (`process._linkedBinding` missing) and to swallow the
+    // "No such binding was linked" error — exactly the Linux fallback this
+    // patch injected. When that native-safe shape is present there is nothing
+    // to patch, so stand down silently instead of failing the required patch.
+    const upstreamReturnsNullOnMissingBinding =
+      /let ([A-Za-z_$][\w$]*)=process\._linkedBinding;if\(typeof \1!=`function`\)return null;/u.test(
+        currentSource,
+      );
+    if (
+      upstreamReturnsNullOnMissingBinding &&
+      currentSource.includes("No such binding was linked:")
+    ) {
+      return currentSource;
+    }
     console.warn(
       "WARN: Could not find Owl feature binding loader - skipping Linux Owl feature fallback patch",
     );

--- a/scripts/patches/main-process/window.js
+++ b/scripts/patches/main-process/window.js
@@ -511,6 +511,23 @@ process.platform===\`linux\`?null:$1?$2($3):null,
 process.platform===\`linux\`?Promise.resolve((()=>{let __codexLinuxAboutIcon=$4.nativeImage.createFromPath(${iconPathExpression});return __codexLinuxAboutIcon.isEmpty()?null:__codexLinuxAboutIcon})()):$4.app.getFileIcon($5,{size:process.platform===\`win32\`?\`large\`:\`normal\`}).catch(()=>null)
 ]`,
     );
+    if (patchedSource === currentSource) {
+      // 26.623 reshaped the about icon promise array: the non-win32 size
+      // ternary collapsed to {size:`normal`} and a win32 nativeImage branch was
+      // added — [t?k_(i):null,n?a.nativeImage.createFromPath(i):a.app.getFileIcon(i,{size:`normal`})].
+      // Without this branch the Linux-safe icon (and the .catch on getFileIcon)
+      // never apply, so a getFileIcon rejection on Linux makes the About window
+      // builder throw before its try/catch and the dialog never opens.
+      const aboutIconPromiseRegex26623 =
+        /\[([A-Za-z_$][\w$]*)\?([A-Za-z_$][\w$]*)\(([^()]+)\):null,([A-Za-z_$][\w$]*)\?([A-Za-z_$][\w$]*)\.nativeImage\.createFromPath\(([^()]+)\):([A-Za-z_$][\w$]*)\.app\.getFileIcon\(([^()]+),\{size:`normal`\}\)\]/;
+      patchedSource = patchedSource.replace(
+        aboutIconPromiseRegex26623,
+        `[
+process.platform===\`linux\`?null:$1?$2($3):null,
+process.platform===\`linux\`?Promise.resolve((()=>{let __codexLinuxAboutIcon=$5.nativeImage.createFromPath(${iconPathExpression});return __codexLinuxAboutIcon.isEmpty()?null:__codexLinuxAboutIcon})()):$4?$5.nativeImage.createFromPath($6):$7.app.getFileIcon($8,{size:\`normal\`}).catch(()=>null)
+]`,
+      );
+    }
   } else {
     const patchedGetFileIconRegex =
       /([A-Za-z_$][\w$]*)\.app\.getFileIcon\(([^()]+),\{size:process\.platform===`win32`\?`large`:`normal`\}\)\.catch\(\(\)=>null\)/;
@@ -521,6 +538,20 @@ process.platform===\`linux\`?Promise.resolve((()=>{let __codexLinuxAboutIcon=$4.
         getFileIconRegex,
         "$1.app.getFileIcon($2,{size:process.platform===`win32`?`large`:`normal`}).catch(()=>null)",
       );
+    }
+    if (patchedSource === currentSource) {
+      // 26.623 fallback (no bundled icon): just make the reshaped getFileIcon
+      // call rejection-proof so the About window builder cannot throw on Linux.
+      const patchedGetFileIconRegex26623 =
+        /([A-Za-z_$][\w$]*)\.app\.getFileIcon\(([^()]+),\{size:`normal`\}\)\.catch\(\(\)=>null\)/;
+      if (!patchedGetFileIconRegex26623.test(patchedSource)) {
+        const getFileIconRegex26623 =
+          /([A-Za-z_$][\w$]*)\.app\.getFileIcon\(([^()]+),\{size:`normal`\}\)/;
+        patchedSource = patchedSource.replace(
+          getFileIconRegex26623,
+          "$1.app.getFileIcon($2,{size:`normal`}).catch(()=>null)",
+        );
+      }
     }
   }
 

--- a/scripts/patches/webview-assets.js
+++ b/scripts/patches/webview-assets.js
@@ -545,7 +545,7 @@ function applyLinuxBrowserUseExternalAvailabilityPatch(currentSource) {
   const availabilityPattern =
     /let ([A-Za-z_$][\w$]*)=([A-Za-z_$][\w$]*)\(([A-Za-z_$][\w$]*)\),([A-Za-z_$][\w$]*)=([A-Za-z_$][\w$]*)===`chrome-extension`\|\|([A-Za-z_$][\w$]*)&&\1\.enabled&&!\1\.isLoading,([A-Za-z_$][\w$]*)=\5===`chrome-extension`\?!1:\1\.isLoading,/g;
 
-  const patchedSource = currentSource.replace(
+  let patchedSource = currentSource.replace(
     availabilityPattern,
     (
       match,
@@ -568,6 +568,22 @@ function applyLinuxBrowserUseExternalAvailabilityPatch(currentSource) {
       return `let ${featureQueryVar}=${featureQueryFn}(${featureQueryArg}),${availableVar}=${windowTypeVar}===\`chrome-extension\`||navigator.userAgent.includes(\`Linux\`)||${statsigVar}&&${featureQueryVar}.enabled&&!${featureQueryVar}.isLoading,${loadingVar}=${windowTypeVar}===\`chrome-extension\`||navigator.userAgent.includes(\`Linux\`)?!1:${featureQueryVar}.isLoading,`;
     },
   );
+
+  if (!changed) {
+    // 26.623 refactored the inline availability gate into a status-string helper:
+    //   function X({isExternalBrowserUseFeatureEnabled:e,isExternalBrowserUseFeatureLoading:t,
+    //     isExternalBrowserUseGateEnabled:n,windowType:r}){return r===`chrome-extension`?`available`:...}
+    // Treat Linux like chrome-extension so the resolved status is `available`.
+    const statusFnPattern =
+      /(function [A-Za-z_$][\w$]*\(\{isExternalBrowserUseFeatureEnabled:[A-Za-z_$][\w$]*,isExternalBrowserUseFeatureLoading:[A-Za-z_$][\w$]*,isExternalBrowserUseGateEnabled:[A-Za-z_$][\w$]*,windowType:([A-Za-z_$][\w$]*)\}\)\{return )\2===`chrome-extension`\?`available`:/;
+    patchedSource = patchedSource.replace(
+      statusFnPattern,
+      (match, prefix, windowTypeVar) => {
+        changed = true;
+        return `${prefix}${windowTypeVar}===\`chrome-extension\`||navigator.userAgent.includes(\`Linux\`)?\`available\`:`;
+      },
+    );
+  }
 
   if (changed || alreadyPatched()) {
     return patchedSource;
@@ -1018,9 +1034,11 @@ function applySubagentNicknameMetadataPatch(currentSource) {
   if (
     patchedSource === currentSource &&
     !(sourceShapePatchedRegex.test(currentSource) && nicknamePatchedRegex.test(currentSource)) &&
-    (currentSource.includes("agentNickname") ||
-      currentSource.includes("agent_nickname") ||
-      currentSource.includes("thread_spawn"))
+    // `thread_spawn` uniquely marks the subagent metadata module. Other webview
+    // chunks reference `agentNickname` without carrying these needles, so gate
+    // the warning on `thread_spawn` to avoid false drift alarms when the patch
+    // pattern matches the shared bundle alongside unrelated chunks.
+    currentSource.includes("thread_spawn")
   ) {
     console.warn("WARN: Could not find subagent nickname metadata needles — skipping metadata shape patch");
   }
@@ -1231,6 +1249,13 @@ function applyBrowserAnnotationScreenshotPatch(currentSource) {
       "Ze=(qe?A?.kind===`comment`?ge:[]:Xe==null?ge:ge.filter(e=>e.id!==Xe.id)).flatMap";
     const electron42CommentPreloadMarkersPatch =
       "Ze=(qe?A?.kind===`comment`?Ke:[]:Xe==null?ge:ge.filter(e=>e.id!==Xe.id)).flatMap";
+    // 26.623 refactored the marker-list computation into imperative form and
+    // adopted the screenshot fix natively: when a comment is selected it now
+    // assigns `it=rt?[j.annotation]:ye`, i.e. only the selected comment's marker
+    // is shown in screenshot mode. Detect that native-safe shape so we skip the
+    // patch without warning.
+    const nativeCommentPreloadMarkersRegex =
+      /([A-Za-z_$][\w$]*)\?\.kind===`comment`\?([A-Za-z_$][\w$]*)=([A-Za-z_$][\w$]*)\?\[\1\.annotation\]:([A-Za-z_$][\w$]*):\3\|\|[A-Za-z_$][\w$]*\?\2=\[\]:[A-Za-z_$][\w$]*!=null&&\(\2=\4\.filter\(e=>e\.id!==[A-Za-z_$][\w$]*\.id\)\)/;
     if (patchedSource.includes(currentMarkersPatch)) {
       // Already patched.
     } else if (patchedSource.includes(currentSelectedMarkersPatch)) {
@@ -1267,6 +1292,9 @@ function applyBrowserAnnotationScreenshotPatch(currentSource) {
         electron42CommentPreloadMarkersNeedle,
         electron42CommentPreloadMarkersPatch,
       );
+    } else if (nativeCommentPreloadMarkersRegex.test(patchedSource)) {
+      // Already native: upstream now scopes screenshot markers to the selected
+      // comment, so no marker patch is required for this build.
     } else {
       console.warn("WARN: Could not find browser annotation screenshot markers — skipping screenshot marker patch");
     }

